// @formatjs/intl-utils/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var diff_1 = require("./diff");
exports.selectUnit = diff_1.selectUnit;
var polyfill_utils_1 = require("./polyfill-utils");
exports.defaultNumberOption = polyfill_utils_1.defaultNumberOption;
exports.getInternalSlot = polyfill_utils_1.getInternalSlot;
exports.getMultiInternalSlots = polyfill_utils_1.getMultiInternalSlots;
exports.getNumberOption = polyfill_utils_1.getNumberOption;
exports.getOption = polyfill_utils_1.getOption;
exports.isLiteralPart = polyfill_utils_1.isLiteralPart;
exports.partitionPattern = polyfill_utils_1.partitionPattern;
exports.setInternalSlot = polyfill_utils_1.setInternalSlot;
exports.setMultiInternalSlots = polyfill_utils_1.setMultiInternalSlots;
exports.setNumberFormatDigitOptions = polyfill_utils_1.setNumberFormatDigitOptions;
exports.toObject = polyfill_utils_1.toObject;
exports.objectIs = polyfill_utils_1.objectIs;
exports.isWellFormedCurrencyCode = polyfill_utils_1.isWellFormedCurrencyCode;
exports.toString = polyfill_utils_1.toString;
exports.formatNumericToString = polyfill_utils_1.formatNumericToString;
exports.toRawFixed = polyfill_utils_1.toRawFixed;
exports.toRawPrecision = polyfill_utils_1.toRawPrecision;
exports.getMagnitude = polyfill_utils_1.getMagnitude;
exports.repeat = polyfill_utils_1.repeat;
exports.hasOwnProperty = polyfill_utils_1.hasOwnProperty;
exports.isWellFormedUnitIdentifier = polyfill_utils_1.isWellFormedUnitIdentifier;
exports.defineProperty = polyfill_utils_1.defineProperty;
var resolve_locale_1 = require("./resolve-locale");
exports.createResolveLocale = resolve_locale_1.createResolveLocale;
exports.getLocaleHierarchy = resolve_locale_1.getLocaleHierarchy;
exports.supportedLocales = resolve_locale_1.supportedLocales;
exports.unpackData = resolve_locale_1.unpackData;
exports.isMissingLocaleDataError = resolve_locale_1.isMissingLocaleDataError;
__export(require("./units"));
var units_1 = require("./units");
exports.removeUnitNamespace = units_1.removeUnitNamespace;
var invariant_1 = require("./invariant");
exports.invariant = invariant_1.invariant;
